### DISCUSSION TITLE
How to handle asynchronous operations with `async`/`await`?

### DISCUSSION THREAD

**Discussion by anujgaurave** [nodejs/help#4638] [Q&A] [answered]
### JavaScript: How to handle asynchronous operations with `async`/`await`?

Hi everyone,

I have been working with asynchronous JavaScript and using `async`/`await` for handling promises. However, I am running into some issues when I have multiple asynchronous operations that depend on one another.

Could someone explain the best way to handle these scenarios? For instance, how can I ensure that one asynchronous operation waits for another to complete? Here's a simplified example of what I am trying to achieve:

```js
async function fetchData() {
  try {
    const response1 = await fetch('https://api.example.com/data1');
    const data1 = await response1.json();

    const response2 = await fetch(`https://api.example.com/data2/${data1.id}`);
    const data2 = await response2.json();

    console.log(data2);
  } catch (error) {
    console.error('Error fetching data:', error);
  }
}

fetchData();

**Comment by Sachin2815** [comment, marked as answer]
### Handling Asynchronous Operations with `async`/`await` in JavaScript

Working with asynchronous operations in JavaScript using `async`/`await` can dramatically simplify your code and make it more readable. Especially when dealing with operations that depend on one another, `async`/`await` helps avoid the dreaded "callback hell" or excessive `.then()` chaining, while keeping the logic flow intuitive and clean.

### Understanding `async`/`await`

In your case, you're trying to:

1. Fetch data from one API endpoint.
2. Use the result of that first fetch to query a second API.

This is a classic case where `async`/`await` shines because it allows you to write asynchronous code in a synchronous style, making the dependency between operations much clearer.

### Solution Explanation

Here’s how you can structure your code:

```javascript
async function fetchData() {
  try {
    // Step 1: Fetch the first piece of data
    const response1 = await fetch('https://api.example.com/data1');
    const data1 = await response1.json();

    // Step 2: Use the result from data1 to fetch the second resource
    const response2 = await fetch(`https://api.example.com/data2/${data1.id}`);
    const data2 = await response2.json();

    // Output the final result
    console.log(data2);
  } catch (error) {
    // In case of any error during the fetch process
    console.error('Error fetching data:', error);
  }
}

fetchData();
